keps/sig-node/3673-kubelet-parallel-image-pull-limit/README.md

```diff
@@ -0,0 +1,862 @@
+
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+* Limiting the number of image layers being downloaded at the same time.
+* Prioritizing image pulls in any way.
+* Using the number of inflight image pulls as a signal to direct pod scheduling.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+* A new integer field `maxParallelImagePulls` will be added to Kubelet Configuration. `maxParallelImagePulls` is the maximum number of inflight image pulls.
+* The default value of `maxParallelImagePulls` will be set to 0, which applies no limit on parallel pulls, the same as current behavior.
+* When `serialize-image-pulls` is true, `maxParallelImagePulls` will be ignored. A warning message will be logged by kubelet if `serialize-image-pulls` is true and `maxParallelImagePulls` is larger than 1.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+A kubernetes user wants to enable parallel image pulls to reduce workload startup latency, but they don't want too many images to be pulled at the same time, as that might burden the disk IO and therefore has a negative impact on disk performance, or might use too much network bandwidth. By setting a proper `maxParallelImagePulls`, they will have a better control over the parallelism and therefore avoid overwhelming the disk.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Since this feature is purely in-memory, if kubelet restarts, it will lose track of the image pulls before the restart and therefore allow the next N (N is the value of `maxParallelImagePulls`) image pulls to proceed.
+In worst case, there could be 2*N image pulls in parallel, if kubelet restarts in the middle.
```

Comment:
Well - technically there can be many more if kubelet is in some crashloop or sth.

So let's change to mention that there can be N+M, where M are pulls that didn't yet finish from previous kubelet runs (need rephrasing to make it clear).

Reply:
I did a simple test to stop kubelet, the pull will be canceled when I stop kubelet.
I wondered if there will be N+M image pulls in parallel.

M may be previous pulls or hanging pulls(not cancelled).

> 17:14:35 paco-centos-9 containerd[4034779]: time="2023-02-03T17:14:35.904425914+08:00" level=error msg="PullImage \"registry.k8s.io/build-image/kube-cross:v1.26.0-go1.19.4-bullseye.1\" failed" error="rpc error: code = Canceled desc = failed to pull and unpack image \"registry.k8s.io/build-image/kube-cross:v1.26.0-go1.19.4-bullseye.1\": failed to resolve reference \"registry.k8s.io/build-image/kube-cross:v1.26.0-go1.19.4-bullseye.1\": failed to do request: context canceled"

Reply:
> I did a simple test to stop kubelet, the pull will be canceled when I stop kubelet.

I guess that was graceful stop, right? What about crash or sth like that?

Reply:
I used containerd v1.6.15 with almost the latest code in the master branch.
Even `kill -9 kubelet` ends up with a `context canceled` on the containerd side.

Reply:
OK - that's interesting.

So let's maybe switch the text to describe it. Something like:
`
Since this feature is purely in-memory, if kubelet restarts, it will lose track of the image pulls before the Kubelet restart. It seems that even on non-graceful shutdown of Kubelet in-flight image pulls are cancelled, but there might be some corner cases when previous image-pulls may still be performed after restart, exceeding the total of `maxParallelImagePulls`.
`

How is that?

Reply:
this is done

Reply:
yep sounds good. Thanks for the suggestions!

Reply:
> I did a simple test to stop kubelet, the pull will be canceled when I stop kubelet. I wondered if there will be N+M image pulls in parallel.
> 
> M may be previous pulls or hanging pulls(not cancelled).
> 
> > 17:14:35 paco-centos-9 containerd[4034779]: time="2023-02-03T17:14:35.904425914+08:00" level=error msg="PullImage "registry.k8s.io/build-image/kube-cross:v1.26.0-go1.19.4-bullseye.1" failed" error="rpc error: code = Canceled desc = failed to pull and unpack image "registry.k8s.io/build-image/kube-cross:v1.26.0-go1.19.4-bullseye.1": failed to resolve reference "registry.k8s.io/build-image/kube-cross:v1.26.0-go1.19.4-bullseye.1": failed to do request: context canceled"

Thanks for testing this case!